### ISSUE TITLE
ambiguous-chained-operator

### ISSUE THREAD

**Issue by kapkekes** [astral-sh/ruff#20531]
### Context

A few days ago, I stumbled upon this snippet:

```python
False == False in [False]
```

At first, I thought that it evaluates to `False`, *which is wrong*.

That day, I learned that `in` and `not in` are actually [comparison operators](https://docs.python.org/3/reference/expressions.html#comparisons), which makes them chainable.

You can argue that this snippet is a little unobvious, but not terrible; so, I modified it:

```python
False == False in [False] == [False]
```

This lil' guy also evaluates to `True`, but in my opinion, it doesn't even look like valid Python code.

### Proposal

My feeling is that the knowledge of `in` being a comparison (so chainable) operator is more of a secret, and not something obvious. I want to suggest a rule that forbids using `in` in chained comparisons.

I think, the samples I provided are not typical code anyone would write. Strictly speaking, they are the first examples of chained `in` I have ever seen. However, if the proposed rule makes sense, I am willing to implement it.

Note that the [PLR1716](https://docs.astral.sh/ruff/rules/boolean-chained-comparison/#boolean-chained-comparison-plr1716) rule may conflict with this one, I haven't checked for that yet.

### PS

While writing this proposal, I considered another rule that would prohibit things like `a < b >= c`; more formal definition would be "if a chained comparison contains `<=` or `<`, it should only use `<=` and `<` (same with `>` and `>=`)", so:

- `a < b <= c <= d` - valid;
- `a > b > c >= d` - valid;
- `a < b >= c == d` - invalid.

I would like to hear feedback on this suggestion, thanks in advance.

**Comment by ntBre** [comment]
This is interesting and does look confusing to me. It looks like PLR1716 only checks for chained less than and greater than operators, much like your PS suggestion, so it shouldn't conflict, good idea to check though!

**Comment by kapkekes** [comment]
I'll start implementing both proposals this weekend as separate rules: `chained-in-usage` and `ambiguous-chained-operator`; I'm too interested to not try.

**Comment by adavis444** [comment]
Although [PLR1716](https://docs.astral.sh/ruff/rules/boolean-chained-comparison/#boolean-chained-comparison-plr1716) comes from Pylint, Pylint has a related warning: [bad-chained-comparison / W3601](https://pylint.readthedocs.io/en/stable/user_guide/messages/warning/bad-chained-comparison.html)

From the tracker in https://github.com/astral-sh/ruff/issues/970, there is no Ruff equivalent.
It would be great to see it implemented!

**Comment by tjkuson** [comment]
Should this issue be closed in favour of #8964? FWIW, I wrote a PR for this, and it was rejected until #1774 gets resolved, though that might not be the case any more.

**Comment by serjflint** [comment]
I wonder if it can go in the preview category.

**Comment by kapkekes** [comment]
@ntBre, hi!

Can you take a look at the situation?

**Comment by ntBre** [comment]
Thanks @tjkuson, I missed the earlier issue and PR. I think the reasoning related to #1774 still stands (this is an opinionated/stylistic lint that restricts valid Python syntax), so I think we should continue holding off on implementing this and fold this issue into #8964.

Sorry for not catching this earlier @kapkekes. We should be able to pull from both your and @tjkuson's PRs when we are ready to add this rule.
